open-craft/opencraft
# Prevent 500 errors when OpenStack API cannot be reached

Currently, there are some cases where a 500 error will be raised in the API view for individual appservers. This happens as a result of a failure to contact the OpenStack API.

This pull requests adds logic to...

1. Set the server state to "Unknown" when OpenStack cannot be properly reached as part of a status check
2. Catch a broader range of exceptions that will result in same
3. By default, check the status on each API call
4. Add a flag to prevent checking the status on each API call

**Testing instructions**:

1. In a local IM instance, spawn an appserver and wait for it to provision.
2. In a local IM instance, go to the API endpoint for that appserver (http://localhost:5000/api/openedx_appserver/{server_id}).
3. Under the `server` subkey, observe that the status is "ready" or similar.
4. Disconnect the machine you're hosting the IM on from the Internet, and refresh the endpoint.
5. Observe that the status of the `server` object has changed to "unknown".
6. Reconnect to the Internet again, and refresh the endpoint a final time
7. Observe that the status has changed back to "ready".

**Author notes and concerns**:

1. Right now, servers don't seem to get updated statii proactively on any given schedule; as a result, I added status fetching to the serializer so that we're returning up-to-date information in the API.

2. It was necessary to catch broad exceptions because the Nova client doesn't appear to comprehensively catch and wrap lower-level exceptions in its own format. And really, we're interested in catching anything that prevents data from being returned from OpenStack.

**Reviewers**
